crates/codegen/syntax_templates/src/shared/cst.rs

```diff
@@ -1,167 +1,62 @@
-        }
-
-        return Rc::new(Node::Token {
+        return Self::Rule(Rc::new(RuleNode {
             kind,
-            range,
-            trivia,
-        });
+            text_len,
+            children: children,
+        }));
     }
-}
 
-pub(crate) enum NodeBuilder {
-    Empty { position: TextPosition },
-    Single { node: Rc<Node> },
-    Multiple { nodes: Vec<Rc<Node>> },
-}
-
-#[allow(dead_code)]
-impl NodeBuilder {
-    pub(crate) fn empty(position: TextPosition) -> Self {
-        return Self::Empty { position };
-    }
-
-    pub(crate) fn single(node: Rc<Node>) -> Self {
-        return Self::Single { node };
+    #[allow(dead_code)]
+    pub(crate) fn token(kind: TokenKind, text: String) -> Self {
+        Self::Token(Rc::new(TokenNode { kind, text }))
     }
 
-    pub(crate) fn multiple(builders: Vec<Self>) -> Self {
-        assert_ne!(
-            builders.len(),
-            0,
-            "codegen should have used empty() builder instead."
-        );
-
-        if builders.len() == 1 {
-            return builders.into_iter().next().unwrap();
-        }
-
-        let start_position = builders.first().unwrap().range().start;
-
-        let mut nodes = vec![];
-
-        for builder in builders {
-            match builder {
-                Self::Empty { .. } => {}
-                Self::Single { node: other } => nodes.push(other),
-                Self::Multiple { nodes: mut other } => nodes.append(&mut other),
-            }
+    pub fn text_len(&self) -> TextIndex {
+        match self {
+            Self::Rule(node) => node.text_len,
+            Self::Token(node) => (&node.text).into(),
         }
-
-        return if nodes.is_empty() {
-            Self::Empty {
-                position: start_position,
-            }
-        } else {
-            Self::Multiple { nodes }
-        };
-    }
-
-    pub(crate) fn with_kind(self, kind: RuleKind) -> Self {
-        let range = self.range();
-
-        let children = match self {
-            Self::Empty { .. } => vec![],
-            Self::Single { node } => vec![node],
-            Self::Multiple { nodes } => nodes,
-        };
-
-        return Self::Single {
-            node: Node::rule(kind, range, children),
-        };
     }
 
-    pub(crate) fn build(self) -> Rc<Node> {
-        return match self {
-            Self::Single { node } => node,
-            _ => panic!("cannot build unnamed nodes"),
-        };
-    }
-
-    pub(crate) fn range(&self) -> TextRange {
-        return match self {
-            Self::Empty { position } => position.to_owned()..position.to_owned(),
-            Self::Single { node } => node.range_including_trivia(),
-            Self::Multiple { nodes } => Range {
-                start: nodes.first().unwrap().range_including_trivia().start,
-                end: nodes.last().unwrap().range_including_trivia().end,
-            },
-        };
+    pub fn utf8_len(&self) -> usize {
+        match self {
```

Comment:
Is this intended to be part of the public API? I wonder what does it offer over `node.text_len().utf8`?

Reply:
Should we remove `utf8_len` now we have `TextRangeExtensions`?

Reply:
Will do